File: leetcode-go/543.go
package main

import "math"

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func findFurthestPath(root *TreeNode, maxPath *int) int {
	if root == nil {
		// nothing
		return 0
	}
	if root.Left == nil && root.Right == nil {
		// Leaf
		return 1
	}
	furthestPathLeft := findFurthestPath(root.Left, maxPath)
	furthestPathRight := findFurthestPath(root.Right, maxPath)
	if furthestPathLeft+furthestPathRight > *maxPath {
		*maxPath = furthestPathLeft + furthestPathRight
	}
	return int(math.Max(float64(furthestPathLeft), float64(furthestPathRight))) + 1
}

func diameterOfBinaryTree(root *TreeNode) int {
	maxPath := 0
	findFurthestPath(root, &maxPath)
	return maxPath
}

File: leetcode-js/543.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var diameterOfBinaryTree = function (root) {
    if (!root) return 0;
    let result = 0;

    const diameter = (node) => {
        if (!node) return 0;
        let leftPath = diameter(node.left);
        let rightPath = diameter(node.right);
        result = Math.max(result, leftPath + rightPath);
        return Math.max(leftPath, rightPath) + 1;
    };

    diameter(root);
    return result;
};